subsys/dult/motion_detector.c

```diff
@@ -0,0 +1,346 @@
+/*
+ * Copyright (c) 2024 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+ */
+
+#include <stdbool.h>
+#include <errno.h>
+
+#include <zephyr/kernel.h>
+#include <zephyr/random/random.h>
+
+#include "dult.h"
+#include "dult_user.h"
+#include "dult_motion_detector.h"
+
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(dult_motion_detector, CONFIG_DULT_LOG_LEVEL);
+
+#define SEPARATED_UT_SAMPLING_RATE1	K_SECONDS(10)
+#define SEPARATED_UT_SAMPLING_RATE2	K_MSEC(500)
+#define SEPARATED_UT_BACKOFF_PERIOD	K_HOURS(6)
+
+#define SEPARATED_UT_TIMEOUT_PERIOD_MIN		(8 * 60)
+#define SEPARATED_UT_TIMEOUT_PERIOD_MAX		(24 * 60)
+#define SEPARATED_UT_TIMEOUT_PERIOD_DIFF	\
+	(SEPARATED_UT_TIMEOUT_PERIOD_MAX - SEPARATED_UT_TIMEOUT_PERIOD_MIN)
+
+#define SEPARATED_UT_ACTIVE_POLL_DURATION	K_SECONDS(20)
+#define SEPARATED_UT_MAX_SOUND_COUNT		10
+
+#define TIMER_PERIOD_IS_EQUAL(_running_timer_period, _ref_period) \
+	((_running_timer_period.ticks) == (_ref_period.ticks))
+
+static bool is_enabled;
+static const struct dult_motion_detector_cb *motion_detector_cb;
+static const struct dult_motion_detector_sound_cb *sound_cb;
+
+static void motion_enable_work_handle(struct k_work *work);
+static void motion_poll_work_handle(struct k_work *work);
+static void motion_poll_duration_work_handle(struct k_work *work);
+
+static K_WORK_DELAYABLE_DEFINE(motion_enable_work, motion_enable_work_handle);
+static K_WORK_DELAYABLE_DEFINE(motion_poll_work, motion_poll_work_handle);
+static K_WORK_DELAYABLE_DEFINE(motion_poll_duration_work, motion_poll_duration_work_handle);
+
+enum motion_poll_state {
+	MOTION_POLL_STATE_STOPPED,
+	MOTION_POLL_STATE_PASSIVE,
+	MOTION_POLL_STATE_PASSIVE_SOUND_REQUESTED,
+	MOTION_POLL_STATE_ACTIVE,
+	MOTION_POLL_STATE_ACTIVE_SOUND_REQUESTED,
```

Comment:
shouldn't we indicate all motion detector states?

- UT timeout (separated and before the first poll state)
- backoff (separated and not in the Poll State)
- Idle (near-owner state)

Reply:
These are polling states not module states. I don't want to add additional unused states.

Reply:
Yes, I suggested tracking the overall module state and expanding the current state machine (`motion_poll_state` -> `motion_state`). It is easier to check the overall module state.

However, I won't insist on doing that.

Reply:
Let's leave it as is. I like the current approach more.